src/Database/Query/InsertQuery.php

```diff
@@ -0,0 +1,121 @@
+<?php
+declare(strict_types=1);
+
+/**
+ * CakePHP(tm) : Rapid Development Framework (https://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         5.0.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Database\Query;
+
+use Cake\Database\Exception\DatabaseException;
+use Cake\Database\Expression\ValuesExpression;
+use Cake\Database\Query;
+use InvalidArgumentException;
+
+/**
+ * This class is used to generate INSERT queries for the relational database.
+ */
+class InsertQuery extends Query
+{
+    /**
+     * List of SQL parts that will be used to build this query.
+     *
+     * @var array<string, mixed>
+     */
+    protected array $_parts = [
+        'with' => [],
+        'insert' => [],
+        'modifier' => [],
+        'values' => [],
+        'epilog' => null,
+    ];
+
+    /**
+     * Create an insert query.
+     *
+     * Note calling this method will reset any data previously set
+     * with Query::values().
+     *
+     * @param array $columns The columns to insert into.
+     * @param array<string, string> $types A map between columns & their datatypes.
+     * @return $this
+     * @throws \InvalidArgumentException When there are 0 columns.
+     */
+    public function insert(array $columns, array $types = [])
+    {
+        if (empty($columns)) {
+            throw new InvalidArgumentException('At least 1 column is required to perform an insert.');
+        }
+        $this->_dirty();
+        $this->_type = self::TYPE_INSERT;
```

Comment:
Shouldn't this be set on construction?

Reply:
I plan to drop `$this->_type` and use `instanceof` of checks instead in follow up PRs. Hence didn't bother with it for now.